TMEDIA-747 : Small Manual Promo Block

ecarlisle opened the pull request:
## Description

1. `small-manual-promo-block` has been migrated to the Themes 2.0 architecture.
2. Testing was updated to use React Testing Library.
3. Readme was updated.

## Jira Ticket

- [TMEDIA-747 : Small Manual Promo Block](https://arcpublishing.atlassian.net/browse/TMEDIA-747)

## Acceptance Criteria

No special notes in Jira ticket. Followed [checklist](https://arcpublishing.atlassian.net/wiki/spaces/TI/pages/3106243821/Converting+a+v1+block+to+v2+architecture) for migrating blocks from v1 to v2.

## Test Steps

### Blocks repository and testing with Storybook.
1. Checkout this branch `git checkout TMEDIA-747-small-manual-promo-block`
2. Run storybook with `npm run storybook` and verify functionality corresponding with stories.
    a. Note: Stories displaying different image aspect ratios should be disregarded, pending future work with Image Resizer.

### Block Appearance
(as seen with Firefox grid view)

#### Image Position: Left
<img width="886" alt="image" src="https://user-images.githubusercontent.com/452134/175136488-bd4df880-91a4-4887-b42b-807e4730a484.png">

#### Image Position: Right
<img width="888" alt="image" src="https://user-images.githubusercontent.com/452134/175136809-05a8175c-e9b1-4765-b27f-2d95bef1a729.png">

#### Image Position: Above
<img width="886" alt="image" src="https://user-images.githubusercontent.com/452134/175137097-0d9fa2b5-8e46-4068-aacd-eaa5603c8185.png">

#### Image Position: Below
<img width="885" alt="image" src="https://user-images.githubusercontent.com/452134/175137794-6812449e-3b20-46e2-a885-46a7d346ac97.png">

#### Fallback Image
The `arcSite` fallback image is used if `showImage` is selected but the `imageURL` custom field is a falsy value.
<img width="881" alt="image" src="https://user-images.githubusercontent.com/452134/175697268-e1ad7fc7-3d19-4e45-8954-4000fee29f77.png">
 
### Testing with Feature Pack repository
1. Change the branch of your local `arc-themes-feature-pack` repository to `TMEDIA-747-small-manual-promo-block`.
2. Set `THEMES_BLOCKS_REPO` in the `.env` file to point to your local `arc-themes-blocks` repository.
4. Start fusion while linking to local block with `npx fusion start -f -l @wpmedia/small-manual-promo-block`
5. Verify cosmetic and functional parity.

## Other Notes
1. When horizontal layout is used on the block, CSS Grid was used to maintain the space occupied the the Heading and MediaItem. The Heading area should always have twice the horizontal space of the Image.
2. The [Figma card for small blocks](https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=3507%3A32495) was updated to show the gap difference in horizontal, vertical presentation.
3. Respective PR in `arc-themes-feature-pack` can be found at https://github.com/WPMedia/arc-themes-feature-pack/pull/281 .

## Dependencies or Side Effects

1. `small-manual-promo-block` still uses `@wpmedia/engine-theme-sdk` for `LazyLoad`.

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [X] Confirmed all the test steps a reviewer will follow above are working.
- [X] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [X] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [X] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [X] Confirmed this PR has unit test files
  - [X] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [X] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

ecarlisle committed:
Remove unneeded index file.

ecarlisle committed:
Remove unneeded lines from .npmignore.

ecarlisle committed:
Edits to package.json for block.

ecarlisle committed:
V2 conversion of block.

ecarlisle committed:
V2 conversion of block tests.

ecarlisle committed:
Add new block Sass file.

ecarlisle committed:
Add small-manual-promo-block tokens to storybook Sass.

ecarlisle committed:
Add image ratio config to Storybook story.

ecarlisle committed:
Remove unneeded objects from test file.

ecarlisle committed:
Flex basis fix for small-manual-promo-block w/ horizontal layout.

ecarlisle committed:
Update readme.

ecarlisle committed:
Solve conflict in news.scss.

JackHowa commented on blocks/small-manual-promo-block/README.md:
```diff
@@ -1,47 +1,36 @@
-| **Prop**                       | **Required** | **Type** | **Description**                                                     |
-| ------------------------------ | ------------ | -------- | ------------------------------------------------------------------- |
-| **customFields.imagePosition** | no           | String   | Determine Image Position within the card. Default to Right Position |
+| **Prop**                       | **Required** | **Type** | **Description**                                                                                       |
```
good work documenting @ecarlisle

ecarlisle replied:
Thank you! 👍🏼

JackHowa reviewed (dismissed):
Looks like the image is growing too wide in the chromatic comparison. In another pr, I tested out some of the image strategies we've used in other places to get the closest parity until we have a better image resizer v2. https://github.com/WPMedia/arc-themes-blocks/pull/1388

There were some other examples of using images in the Numbered List block with the v2 architecture. 

I wonder if it's something to do with flex and growing here with the use of min-content here handling the use of space

<img width="1409" alt="Screen Shot 2022-06-20 at 10 32 55" src="https://user-images.githubusercontent.com/5950956/174572441-5a8b49a7-61a9-48ce-ae8b-3e897f5eafa8.png">

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.test.jsx:
```diff
@@ -1,118 +1,120 @@
-		registerSuccessEvent: () => ({}),
-	})),
+jest.mock("@wpmedia/arc-themes-components", () => ({
+	formatURL: jest.fn(),
```
We should not be mocking all these components from Arc Themes Components - The only one we should need to mock is the `isServerSide` function

ecarlisle replied:
Changed. `isServerSide` is the only item mocked from `@wpmedia\arc-themes-components`.

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.test.jsx:
```diff
@@ -1,118 +1,120 @@
 	});
 
+	/*
 	it("should have 1 container fluid class", () => {
```
If test is no longer valid we can delete the code

ecarlisle replied:
I thought I got them all but must have missed these.
Got it. Thanks!

matthewroach commented on .storybook/themes/news.scss:
```diff
@@ -1056,6 +1056,26 @@
+						),
+						image: (
+							display: inline,
+							aspect-ratio: attr(aspect-ratio),
```
This is not working as intended - you need to specify the entire attribute name. It is also not working due to the `Image` component not outputting the HTML attribute

```suggestion
							aspect-ratio: attr(data-aspect-ratio),
```

ecarlisle replied:
Confirmed. The `<Image>` component doesn't have that prop or a `...rest` so I can't send it in.
I believe we were using this for testing. I'll confirm that we need it.

nschubach replied:
It should work to maintain the aspect ration as long as one dimension is constrained... but the image currently doesn't pass along "...rest" like the other components (Was going to see if @matthewroach knew if there was a specific reason or it was missed, since he was the one that initially added them)

matthewroach replied:
Looks like the image component wasn't updated with the rest - the only reason I would say it wasn't done was because it's a barebones component until we have resizer v2 logic added - I do not see a reason to not add `...rest` into the component for flexibility

ecarlisle replied:
@matthewroach , @nschubach : I've taken out usage of the `imageRatio` custom field for now pending availability of Image Resizer 2.0.

matthewroach commented on .storybook/themes/news.scss:
```diff
@@ -1056,6 +1056,26 @@
+					gap: var(--global-spacing-6),
+				),
+				small-manual-promo-vertical: (
+					gap: var(--global-spacing-5),
```
Have we called out this discrepancy between the horizontal and vertical gaps in the discrepancies document? This feels like it should be the same gap for both variations

ecarlisle replied:
Both tokens are called out in the Figma specs:
https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=3507%3A32495

I asked Brandon and Kelly about this when developing the block:
https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr?node-id=3507:32495#212870942

I can include this in the Design Discrepancies doc in Confluence. Though it is in Figma, it might be good to note it in the Confluence docs as well.

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.jsx:
```diff
@@ -1,23 +1,81 @@
 		return null;
 	}
 
+	const PromoImage = () =>
```
We have lost the inline ability to use the integrated photo search selector

ecarlisle replied:
Confirming. Here's what I'm seeing locally when clicking the icon on the "Image URL" custom field. Am I misunderstanding what has been lost?
<img width="1484" alt="image" src="https://user-images.githubusercontent.com/452134/174657429-27c8ae41-3673-464e-85ae-2855e491bb02.png">

matthewroach replied:
Do you still see the inline button as shown here?
<img width="277" alt="Google Chrome-2022-06-20-19-32 47" src="https://user-images.githubusercontent.com/868127/174660664-510bbe70-c583-4464-b5b0-06238b0074eb.png">

ecarlisle replied:
No. Happen to know the cause or means to resolve? If not I can look into it. Thanks!
<img width="232" alt="image" src="https://user-images.githubusercontent.com/452134/174664155-e619bc7b-7954-4d98-a62c-db8d348ff87f.png">

matthewroach replied:
You need to use `searchableField` it should be in the current promo image

```
const { searchableField } = useEditableContent();
```

ecarlisle replied:
TYVM! Will give this a try!

ecarlisle replied:
@matthewroach : Got it, thanks!

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.jsx:
```diff
@@ -1,23 +1,81 @@
 		<LazyLoad enabled={shouldLazyLoad}>
-			<SmallPromoPresentation {...customFields} />
+			<HeadingSection>
+				<Stack
```
This is an implementation detail - we could use `grid` here to allow developers to specific the column more accurately

ecarlisle replied:
In thinking about this, `<Grid>` may allow a more-consistent layout approach in the different sizes of blocks. The small block has either a vertical or horizontal layout so Flexbox works easily. Larger blocks have more-complex layouts that change at breakpoints so they're probably a better fit for CSS Grid. For that reason, we may want to use Grid here to keep block layout approach consistent. Was that close to what you were thinking? I'm happy to give this a try.

Related: How would you feel about adding an `as` property to `<Grid>` like we have in `<Stack>`?
https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/stack/index.jsx#L18-L20

I like how the `as` allows me to:
1. Give the top-level tag a semantic name (in this case, I'm using `<article>`).
2. Not have to send in an `aria-role` in with the `...rest.`

ecarlisle replied:
@matthewroach : I've changed the top-level component from `<Stack>` to `<Grid>` and have changed styling accordingly.

ecarlisle committed:
Remove unneeded tests.

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.jsx:
```diff
@@ -1,23 +1,81 @@
 import React from "react";
 import PropTypes from "@arc-fusion/prop-types";
 import { useFusionContext } from "fusion:context";
```
We are also missing the bandito support that promo's has, you'll need `useComponentContext`

and the following code within the block

```
const { registerSuccessEvent } = useComponentContext();
```

as well as the `onClick` event

```
onClick={registerSuccessEvent}
```

ecarlisle replied:
Thanks! Will add.

ecarlisle replied:
@matthewroach : Added. Thanks!

ecarlisle committed:
Update block Sass tokens to use CSS Grid.

ecarlisle committed:
Update block Sass token hooks to use CSS Grid.

ecarlisle committed:
Remove unneeded scripts from block package.json.

ecarlisle committed:
Changes to the block:

1. Use <Grid> instead of <Stack>.
2. Add missing integration points for Photo Center.
3. Add missing integration point for Bandito.

ecarlisle committed:
Remove unneded mocks. Add needed mocks. Change test expects accordingly.

ecarlisle committed:
Merge branch 'arc-themes-release-version-2.0.1' into TMEDIA-747-small-manual-promo-block

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.jsx:
```diff
@@ -1,23 +1,75 @@
+			<HeadingSection>
+				<Grid
+					role="article"
+					className={`${BLOCK_CLASS_NAME} ${BLOCK_CLASS_NAME}__${imagePosition}`}
```
Using BEM this is a modifier so should use `--` in the class

```suggestion
					className={`${BLOCK_CLASS_NAME} ${BLOCK_CLASS_NAME}--${imagePosition}`}
```

We would also need to update the Scss file to match this

ecarlisle replied:
@matthewroach : Thanks! Changed in SCSS file as well.

ecarlisle committed:
Update blocks/small-manual-promo-block/features/small-manual-promo/default.jsx

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

ecarlisle committed:
Change BEM __ to --.

ecarlisle committed:
Merge branch 'TMEDIA-747-small-manual-promo-block' of github.com:WPMedia/arc-themes-blocks into TMEDIA-747-small-manual-promo-block

matthewroach commented on blocks/small-manual-promo-block/features/small-manual-promo/default.jsx:
```diff
@@ -1,23 +1,75 @@
-			<SmallPromoPresentation {...customFields} />
+			<HeadingSection>
+				<Grid
+					role="article"
```
Created a PR to update the `Grid` component to have an `as` prop so we can abide by the first rule of ARIA and not use ARIA - https://github.com/WPMedia/arc-themes-components/pull/107

ecarlisle replied:
Thanks! Approved and merged that PR.
Made change here so that `<article>` is the top-level element.

matthewroach reviewed (changes requested):
@ecarlisle - While testing this out in PageBuilder I noticed the placeholder image logic is missing. When the promo block is first placed on the page the initial image that is shown is the placeholder image, with these changes there is no placeholder image and the image is rendering as missing and showing the alt text

ecarlisle committed:
Use 'article' semantic tag on <Grid>

ecarlisle committed:
update package file.

ecarlisle committed:
Use fallback image initially.
